File: backend/scripts/keep_alive.py
#!/usr/bin/env python3
"""
Keep-alive script for Render free tier backend
This script pings the backend every 10 minutes to keep it awake
"""
import requests
import time
import os
from datetime import datetime

# Backend URL
BACKEND_URL = "https://store-6ryk.onrender.com"

def ping_backend():
    """Ping the backend to keep it awake"""
    try:
        response = requests.get(f"{BACKEND_URL}/health", timeout=30)
        if response.status_code == 200:
            print(f"✅ {datetime.now().strftime('%Y-%m-%d %H:%M:%S')} - Backend is awake")
            return True
        else:
            print(f"⚠️ {datetime.now().strftime('%Y-%m-%d %H:%M:%S')} - Backend responded with status {response.status_code}")
            return False
    except requests.exceptions.RequestException as e:
        print(f"❌ {datetime.now().strftime('%Y-%m-%d %H:%M:%S')} - Failed to ping backend: {e}")
        return False

def main():
    """Main keep-alive loop"""
    print(f"🔄 Starting keep-alive service for {BACKEND_URL}")
    print("Press Ctrl+C to stop")
    
    while True:
        ping_backend()
        # Wait 10 minutes (600 seconds) before next ping
        time.sleep(600)

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n🛑 Keep-alive service stopped")

File: frontend/src/services/keepAlive.js
/**
 * Keep-alive service to prevent Render backend from sleeping
 * Pings the backend every 10 minutes to keep it awake
 */

class KeepAliveService {
  constructor() {
    this.intervalId = null;
    this.isActive = false;
    this.pingInterval = 10 * 60 * 1000; // 10 minutes
    this.backendUrl = process.env.REACT_APP_API_BASE_URL || 'https://store-6ryk.onrender.com/api/v1';
  }

  start() {
    if (this.isActive) {
      console.log('🔄 Keep-alive service already running');
      return;
    }

    console.log('🔄 Starting keep-alive service');
    this.isActive = true;
    
    // Ping immediately
    this.pingBackend();
    
    // Set up interval
    this.intervalId = setInterval(() => {
      this.pingBackend();
    }, this.pingInterval);
  }

  stop() {
    if (!this.isActive) {
      console.log('🔄 Keep-alive service not running');
      return;
    }

    console.log('🛑 Stopping keep-alive service');
    this.isActive = false;
    
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
  }

  async pingBackend() {
    try {
      const response = await fetch(`${this.backendUrl.replace('/api/v1', '')}/health`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (response.ok) {
        const data = await response.json();
        console.log(`✅ Keep-alive: Backend is awake (${new Date().toLocaleTimeString()})`);
        return true;
      } else {
        console.warn(`⚠️ Keep-alive: Backend responded with status ${response.status}`);
        return false;
      }
    } catch (error) {
      console.error(`❌ Keep-alive: Failed to ping backend:`, error);
      return false;
    }
  }

  // Method to manually wake up the backend
  async wakeUpBackend() {
    console.log('🔄 Manually waking up backend...');
    const success = await this.pingBackend();
    
    if (success) {
      console.log('✅ Backend is now awake');
    } else {
      console.log('❌ Failed to wake up backend');
    }
    
    return success;
  }
}

// Create singleton instance
const keepAliveService = new KeepAliveService();

export default keepAliveService;

